### ISSUE TITLE
General enhacements

### ISSUE THREAD

**Issue by jpbrocca** [UXDivers/Gorilla-Player-Support-And-Issues#41]
For your convenience I have created this issue to add waw info about suggestions/enhancements.

Thanks in advance!

**Comment by Horrorshow** [comment]
1) Some stability improvement on the server and client. Seems to happen when u start visual studio and work the server alot by connecting both in VS2015 and Android client.

_Make sure the server works on socket timeouts. If you quit your client/server make sure the sockets pick it up and properly shows that the client is disconnected. Make sure the server does not need to be restarted due timeouts._

2) If its possible to really deploy the gorillaplayer without force uninstalling on the devices when switching from mac/win.

_It seems like the the gorilla player wont start after you switch from mac/windows to one other then you have to redeploy from the current OS to make it work_

3) Vs2015 better menu for gorillla player like stick to xml onto the solution explorer.

_Navigating tools->gorilla player->stick to xml/connect etc etc, should be fewer and easier steps with maybe a visual indicators if possible._

4) Often the gorilla player needs to be recompiled and relaunched as it just shuts down (i havent checked log on android why) This is again related to step 2.

5) When gorilla player shines is when you can use 3 devices++ connected to the server and design all realtime. Android s5, nexus7 and Iphone 4s. Problem again is that sometimes the socket stability is somewhat not good enough as it seem to time out and needs to be restarted.
A indicator on the server showing whos connected (Device id) would of been helpful.

**Comment by jpbrocca** [comment]
Thanks @Horrorshow, we will review your great suggestions soon and get back to you.

Thanks again!
JP

**Comment by Horrorshow** [comment]
np! :)

**Comment by WillemMali** [comment]
Something I'd like to see is support for AndroidManifest.xml themes; I currently use Theme.Material.Light but in Gorilla Player everything shows up as black and dark gray.

Here's my AndroidManifest.xml:

```
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android" package="testing.testapp" android:installLocation="auto">
	<uses-sdk android:minSdkVersion="15" />
	<application android:label="TestApp"
               android:theme="@android:style/Theme.Material.Light">
  </application>
</manifest>
```

**Comment by LeoHere** [comment]
Thanks for your feedback @Horrorshow, it is really useful. 

Regarding 2), the only way in which Gorilla Player might be tight to the machine it is deployed from is by the Gorilla server name it will look for by default.  During Gorilla installation a name is generated for your Gorilla host app. Usually the name is something like  "Gorilla on <machine name>". If you deploy the player (out of the box) to a device from Machine1 it will automatically try to connect to "Gorilla on Machine1". But you should always be able to connect to other Gorilla host apps. 

You can always change the default Gorilla host app a player looks for by changing the name passed to Config class in AppDelegate or MainActivity, in the player solution.

Can you elaborate a bit more in the steps required to reproduce 2)?

**Comment by Horrorshow** [comment]
LeoHere i havent researched it to the full extent, but the scenario is when i work on xamarin trial (30days) on one machine (a mac) and deploy it do some codework, later switch to a windows machine i have to recompile and rebuild the player on the device or else it just shuts down before starting up.
Ill have a look at the log what the error says, might be my development environment. Sorry for the vague feedback.

**Comment by omanuke** [comment]
I wanna use my own View class as base class which is derived from ContentView or ContentPage, but they doesn't work on Gorilla. Attached it as sample.

Gorilla allows just default classes like ContentView ,ContentPage,TabbedPage,etc?
If it so, it's better that Gorilla ignore derived class and treate as ContentView or ContentPage.

[CustomeView.txt](https://github.com/UXDivers/Gorilla-Player-Support-And-Issues/files/140166/CustomeView.txt)

**Comment by omanuke** [comment]
sorry,related [issues](https://github.com/UXDivers/Gorilla-Player-Support-And-Issues/issues/12) was written already.

**Comment by jpbrocca** [comment]
Hi @omanuke...no worries, thanks for collaboration! :)

**Comment by Horrorshow** [comment]
Is baseclass supported yet?

**Comment by WillemMali** [comment]
44yö

**Comment by 0rupp** [comment]
1) Support cross platform images that use ImageResourceExtension (Official Xamarin Forms: Working with Images article)
2) The SampleData.json seems to work for .xaml files located in root folder only. I have my files in a subfolder, and can't figure out how to bind the data from the .json file to them. It seems to work fine moving the .xaml file down to the root folder, starting gorilla and moving it back to the subfolder
3) The Gorilla App (Android, haven't tested iOS), disconnects when the screen goes off. Then when waking the phone I need to tap my server, scroll down (annoying!) and re-connect. Maybe the connection could be left open or a auto-reconnect of the last-opened connection could occur on start-up.

**Comment by LeoHere** [comment]
Hi @0rupp,
Regarding your requests:
1) Ok, got it.
2) This should work in Gorilla as it is. I just updated [this](https://github.com/UXDivers/Gorilla-Player-Support/wiki/Working-with-sample-data) wiki page with a more in deep example of how it works.
3) This scroll down issue (yes annoying) is fixed in the next version. We also added an option to avoid the device screen to go off. Gorilla should auto connect by default when discovery mode is used and the server name it is looking for is discovered. If you want to go deeper in diagnosing why it is not auto-connecting in you case please open a separate issue.

**Comment by Mr-Pearce** [comment]
I would like a "fake Toolbar/Menubar".
I have a MasterDetailApp and in the preview the Menu-bar is missing which leads to stuff eaten up by the IOS "Carrier Time Battery" bar or that stuff is text in text in that area.

This is how it looks
![2016-09-16_04-08-05-](https://cloud.githubusercontent.com/assets/13457314/18588782/367729c4-7c28-11e6-934e-d68ecb1512c3.jpg)
This is how it should
![2016-09-16_04-08-52-](https://cloud.githubusercontent.com/assets/13457314/18588786/3b12c4ca-7c28-11e6-91c1-8f983191ea4a.jpg)

**Comment by LeoHere** [comment]
@Mr-Pearce did you tried [this](https://github.com/UXDivers/Gorilla-Player-Support/wiki/Previewing-within-a-Navigation-Page)?

**Comment by Mr-Pearce** [comment]
@LeoHere *FacePalm* ... yes now i did. works perfect!

**Comment by EricWu91** [comment]
Hey there. Firstly...  _awesome_ tool. It really saved me a lot of time.
I'd like to know if there is a way of making Gorilla read the static resources from another, referred library.
See if the following structure is clear enough:

Project A
*
*---> App.cs
*---> App.xaml

Project B
*(reference) Project A.dll
*---> Page1.xaml
*---> Page2.xaml

Pages 1 and 2 contain static resources that are present at App.xaml, but aren't found by Gorilla (presumably) because they are in another dll, and an error is thrown (resource not found). 
It would be awesome if these resources could be considered. 

Thanks in advance!

**Comment by LeoHere** [comment]
@EricWu91 AFAIK, what you want to do is not possible in XF. Usually Project A would reference project B and the ```App.xaml``` would be the XF ```Application```. In this case you will be able to access the resources in the ```App.xaml``` from ```Page1.xaml``` and ```Page2.xaml```.

**Comment by EricWu91** [comment]
Sorry, @LeoHere , I was typing in a hurry and didn't see the actual structure. I updated the comment. It is exactly as you said. The question remains, though. Gorilla doesn't seem to be able to read the resources contained at App.XAML (throws an error).

**Comment by LeoHere** [comment]
It should. Any chance you can share the app source or generate a test proj where we can reproduce the issue?

**Comment by EricWu91** [comment]
Hm, sadly, the codebase is proprietary (and too large) to be shared. :( 
If there are any logs GP generates, I'll be happy to upload them.

**Comment by LeoHere** [comment]
Got it. Ok, set Gorilla to log level debug (see [this](https://github.com/UXDivers/Gorilla-Player-Support/wiki/Gorilla-Player-logs-location)) and send me those logs to l r o d r i g u e z at u x d i v e r s . c o m

**Comment by EricWu91** [comment]
@LeoHere , thanks for the helpfulness. It is really appreciated. 
Truth be told, it actually doesn't make much sense to _work_ under Gorilla.
If Project B (which I call "_Base_") will only have the .xaml files, if I want to see them in GP, I'll have to go to that specific project and try to visualize it. Thing is, the resources are, indeed, in another dll it has no sight on. Sorry to waste your time :(

**Comment by EricWu91** [comment]
As a sidenote though... I seem to be facing issues on loading UXDivers.Gorilla.Common on iOS, both in Simulator and Device. Are there reported issues?

**Comment by aubykhan** [comment]
@LeoHere I have modified the Gorilla Player on Android to support **Material Design** themes. Was looking for a GitHub repo to add a pull request for these changes. Possible? or is it already supported in latest Player?

**Comment by LeoHere** [comment]
@EricWu91, sorry for the delay, which version of XF are your using? (forcing the assembly to be loaded by doing a ```typeof<UXDivers.Gorilla.Platform.JsonLoader>()```) 

@aubykhan, no the current android player is not based on AppCompat. We will consider adding it.   It is not clear for me if most of XF Android projects are indeed based on AppCompat, any info on that? If that is the case it makes lot of sense to move to AppCompat.

**Comment by aubykhan** [comment]
@LeoHere I think most of them are indeed using AppCompat now. Even if you create a new project, it's made as AppCompat by default. 

I have create a repo for that. It works perfectly and uses LightTheme with DarkActionBar:
[https://github.com/aubykhan/Gorilla-Player---Material](https://github.com/aubykhan/Gorilla-Player---Material)

**Comment by LeoHere** [comment]
Ok, cool thanks!!

**Comment by EricWu91** [comment]
@LeoHere , I'm using 0.9.1.2 (the last stable version in Nuget)

**Comment by EricWu91** [comment]
Also, I can confirm @aubykhan in that Xamarin Forms' Android projects are born with a AppCompat Theme.
